### PR TITLE
Support Chunked DPO Loss Kernel

### PR DESCRIPTION
**Pull request by austin362667**
## Summary

Add support for a fused, torch-compiled, and chunked DPO ([Direct Preference Optimization](https://arxiv.org/html/2305.18290v3)) loss kernel, as requested in https://github.com/linkedin/Liger-Kernel/issues/371.
This implementation is largely based on the excellent work done on ORPO (https://github.com/linkedin/Liger-Kernel/pull/362) by @shivam15s.

### DPO Loss Formulation

In a reference setting (not reference free):

$$r_\theta(x,y_c) - r_\theta(x,y_r) = \log(\pi_\theta(y_c|x)) - \log(\pi_\theta(y_r|x))$$

$$-\log(\sigma((\log(\pi_\theta(y_c|x)) - \log(\pi_\theta(y_r|x)) - \log(\pi_{\theta_{\text{ref}}}(y_c|x)) + \log(\pi_{\theta_{\text{ref}}}(y_r|x)))/\beta))$$

Corresponds to:
```python
# Policy model log probabilities
policy_chosen_logps = log_probs(policy_chosen_logits)
policy_rejected_logps = log_probs(policy_rejected_logits)

# Reference model log probabilities
ref_chosen_logps = log_probs(ref_chosen_logits)
ref_rejected_logps = log_probs(ref_rejected_logits)

# Compute advantages
chosen_advantages = policy_chosen_logps - ref_chosen_logps
rejected_advantages = policy_rejected_logps - ref_rejected_logps

# DPO loss
logits_diff = (chosen_advantages - rejected_advantages) / beta
losses = -F.logsigmoid(logits_diff)
```

In this PR:

1. The above mathematical equation shows that to maximize the reward difference, we get formula:
    $$r_θ(x_c) - r_θ(x_r)$$
2. This can be further optimized using just:
    $$-log(σ((π_θ(x_c) - π_θ(x_r))/β))$$
3. So, the code implements:
    ```python
    logits_diff = (chosen_logps - rejected_logps) / beta  # (π_θ(x_c) - π_θ(x_r))/β
    losses = -F.logsigmoid(logits_diff)                   # -log(σ(logits_diff))
    ```
4. Sum up DPO and NLL:
    $$L_{DPO+NLL} = L_{DPO}+αL_{NLL}$$

## Testing Done

dpo_loss_memory
dpo_loss_speed

- Hardware Type: **NVIDIA L40S (48G)**
- [X] run `make test` to ensure correctness
- [X] run `make checkstyle` to ensure code style
- [X] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: benchmark/scripts/benchmark_dpo_loss.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+from utils import (
+    QUANTILES,
+    SingleBenchmarkRunInput,
+    SingleBenchmarkRunOutput,
+    _test_memory,
+    parse_benchmark_script_args,
+    run_benchmarks,
+)
+
+from liger_kernel.alignment.dpo_loss import HF_DPO_Loss, LigerFusedLinearDPOFunction
```
Comment (austin362667) [inline benchmark/scripts/benchmark_dpo_loss.py:12]: Should I use HF DPO impl here in benchmarking for function reusability purpose?

File: src/liger_kernel/alignment/dpo_loss.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                    grad_bias.stride(-1),
+                    grad_output,
+                    1,
+                    BLOCK_SIZE=BLOCK_SIZE,
+                    num_warps=32,
+                )
+        return grad_input, grad_weight, None, grad_bias, None, None, None
+
+
+class HF_DPO_Loss:
```
Comment (austin362667) [inline src/liger_kernel/alignment/dpo_loss.py:210]: Should I move this  HF impl to file `test_dpo_loss.py`?

File: benchmark/scripts/benchmark_dpo_loss.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    mem_50, mem_20, mem_80 = _test_memory(full, _iter=10, quantiles=QUANTILES)
+    return SingleBenchmarkRunOutput(
+        y_20=mem_20,
+        y_50=mem_50,
+        y_80=mem_80,
+    )
+
+
+def bench_speed_dpo_loss(input: SingleBenchmarkRunInput) -> SingleBenchmarkRunOutput:
+    B = input.x // 2
```
Comment (austin362667) [inline benchmark/scripts/benchmark_dpo_loss.py:117]: TO show that we have input shape of pairs of preference dataset? (Perhaps this is unnecessary)

File: src/liger_kernel/alignment/dpo_loss.py (hunk 2 above)
Comment (Tcc0403) [reply src/liger_kernel/alignment/dpo_loss.py:210]: yes, since HF impl is only for testing purpose

**Review by pramodith — COMMENTED** [review]
Just a FYI, I think we should wait until @shivam15s pushes a generic/inheritable class that handles all the chunking and other repetitive logic common to different loss functions, before pushing new loss functions.

**Comment by shivam15s** [comment]
Great work @austin362667 ! The additional summing of NLL loss is going to be useful for [IRPO loss](https://github.com/huggingface/trl/issues/1611) as well :). I'll be creating a simple base class which adds the boilerplate code (backward/torch compile logic) that you can inherit from, as @pramodith mentioned

**Comment by austin362667** [comment]
Issue addressed. Thanks @Tcc0403 @lancerts  @pramodith @shivam15s and @ByronHsu for review!

**Comment by Tcc0403** [comment]
I think we should make chunked_loss functions nn.Module (like flce and fljsd) for users? same for orpo? cc @shivam15s @ByronHsu

**Comment by ByronHsu** [comment]
@Tcc0403 that is the plan!

**Review by ByronHsu — APPROVED** [review]
